File: example/1d-linear-convection/weno3/cpp/02a/src/cfd_solver.cpp
// ==================== src/cfd_solver.cpp (重构版) ====================
#include "cfd_solver.hpp"

namespace cfd {

void CfdSolver::initialize(const CfdConfig& config) {
    config_ = config;
    config_.validate();
    
    // Initialize domain
    domain_ = ComputationalDomain(config_);
    
    // Initialize solution
    solution_ = Solution(domain_);
    
    // Create components using factories
    boundary_condition_ = BoundaryFactory::create_boundary(config_);
    
    // Default: step function for 1D convection
    std::vector<Real> step_params = {1.0, 2.0, 0.5, 1.0};
    initial_condition_ = InitialConditionFactory::create_initial_condition("step", step_params);
    
    // Create residual calculator
    residual_calculator_ = ResidualCalculatorFactory::create_from_config(config_, true); // compact mode
    
    // Create time integrator
    time_integrator_ = TimeIntegratorFactory::create_integrator(
        config_.rk_order,
        std::make_unique<ConvectionResidualCalculator>(
            ReconstructorFactory::create_reconstructor(config_),
            FluxFactory::create_flux_calculator(config_.flux_type),
            BoundaryFactory::create_boundary(config_),
            config_.wave_speed
        )
    );
    
    // Initialize solution
    initialize_solution();
    
    // Adjust time step based on CFL condition
    calculate_dt();
    
    if (config_.verbose) {
        std::cout << "CFD solver initialized:" << std::endl;
        std::cout << "  Time integrator: " << time_integrator_->name() << std::endl;
        std::cout << "  Residual calculator: " << residual_calculator_->name() << std::endl;
        std::cout << "  Boundary condition: " << boundary_condition_->name() << std::endl;
        std::cout << "  Initial condition: " << initial_condition_->name() << std::endl;
        std::cout << "  Time step: " << config_.dt << std::endl;
    }
}

void CfdSolver::initialize_solution() {
    // Apply initial condition
    initial_condition_->initialize_with_ghosts(solution_.u(), domain_);
    
    // Apply boundary condition
    boundary_condition_->apply(solution_.u(), domain_);
    
    // Save as old solution
    solution_.copy_to_old();
}

void CfdSolver::calculate_dt() {
    // CFL condition: dt <= CFL * dx / |wave_speed|
    Real dt_cfl = config_.cfl * domain_.mesh().dx() / std::abs(config_.wave_speed);
    
    if (config_.dt > dt_cfl) {
        if (config_.verbose) {
            std::cout << "Adjusting time step for stability:" << std::endl;
            std::cout << "  Original dt = " << config_.dt << std::endl;
            std::cout << "  CFL dt = " << dt_cfl << std::endl;
        }
        config_.dt = dt_cfl;
        if (config_.verbose) {
            std::cout << "  Using dt = " << config_.dt << std::endl;
        }
    }
}

void CfdSolver::set_boundary_condition(std::unique_ptr<BoundaryCondition> bc) {
    boundary_condition_ = std::move(bc);
    // Update residual calculator with new boundary condition
    if (residual_calculator_) {
        residual_calculator_ = ResidualCalculatorFactory::create_convection_residual(
            std::make_unique<EnoReconstructor>(config_.spatial_order),
            FluxFactory::create_flux_calculator(config_.flux_type),
            std::make_unique<PeriodicBoundary>(),
            config_.wave_speed,
            true
        );
    }
    // Re-apply boundary conditions
    boundary_condition_->apply(solution_.u(), domain_);
}

void CfdSolver::set_boundary_condition(const std::string& bc_type, 
                                      Real left_val, Real right_val) {
    set_boundary_condition(BoundaryFactory::create_boundary(bc_type, left_val, right_val));
}

void CfdSolver::set_initial_condition(std::unique_ptr<InitialCondition> ic) {
    initial_condition_ = std::move(ic);
    initialize_solution();
}

void CfdSolver::set_initial_condition(const std::string& ic_type,
                                     const std::vector<Real>& params) {
    set_initial_condition(InitialConditionFactory::create_initial_condition(ic_type, params));
}

void CfdSolver::set_time_integrator(std::unique_ptr<TimeIntegrator> integrator) {
    time_integrator_ = std::move(integrator);
}

void CfdSolver::set_time_integrator(const std::string& method) {
    time_integrator_ = TimeIntegratorFactory::create_integrator(
        method,
        std::move(residual_calculator_)
    );
}

void CfdSolver::set_residual_calculator(std::unique_ptr<ResidualCalculator> calculator) {
    residual_calculator_ = std::move(calculator);
}

void CfdSolver::advance_one_step(Real dt) {
    if (!time_integrator_) {
        throw std::runtime_error("Time integrator not initialized");
    }
    
    // Perform one time step
    time_integrator_->step(solution_.u(), dt, domain_);
    
    // Save old solution
    solution_.copy_to_old();
}

Vector CfdSolver::run_simulation(Real final_time) {
    // Use config final time if not specified
    if (final_time < 0.0) {
        final_time = config_.final_time;
    }
    
    // Ensure solver is properly initialized
    if (!time_integrator_ || !residual_calculator_ || 
        !boundary_condition_ || !initial_condition_) {
        throw std::runtime_error("Solver not properly initialized. Call initialize() first.");
    }
    
    Real t = 0.0;
    Real dt_old = config_.dt;
    Real dt = dt_old;
    const int max_steps = 100000;  // Safety limit
    
    if (config_.verbose) {
        std::cout << "Starting time integration..." << std::endl;
        std::cout << "  Final time: " << final_time << std::endl;
        std::cout << "  Time step: " << dt << std::endl;
        std::cout << "  CFL number: " << config_.cfl << std::endl;
        std::cout << "  Time integrator: " << time_integrator_->name() << std::endl;
        std::cout << "  Residual calculator: " << residual_calculator_->name() << std::endl;
        std::cout << "  Boundary condition: " << boundary_condition_->name() << std::endl;
        std::cout << "  Initial condition: " << initial_condition_->name() << std::endl;
    }
    
    int step = 0;
    while (t < final_time - 1.0e-12 && step < max_steps) {
        ++step;
        
        if (t + dt > final_time) {
            dt = final_time - t;
        }
        
        advance_one_step(dt);
        t += dt;
        
        // Progress report
        if (config_.verbose && step % config_.print_interval == 0) {
            std::cout << "  Step " << step << ", Time = " << t << std::endl;
        }
    }
    
    if (step >= max_steps) {
        std::cout << "Warning: Reached maximum number of steps (" << max_steps << ")" << std::endl;
    }
    
    config_.dt = dt_old;
    
    if (config_.verbose) {
        std::cout << "Time integration completed:" << std::endl;
        std::cout << "  Total steps: " << step << std::endl;
        std::cout << "  Final time: " << t << std::endl;
    }
    
    // Return interior solution (without ghost cells)
    return solution_.get_interior_solution();
}

Real CfdSolver::analytical_solution(Real x, Real t) const {
    Real L = domain_.mesh().L();
    Real a = config_.wave_speed;
    
    // Periodic boundary handling
    Real x_shifted = x - a * t;
    while (x_shifted < 0.0) x_shifted += L;
    while (x_shifted > L) x_shifted -= L;
    
    // Use the current initial condition to compute analytical solution
    return initial_condition_->evaluate(x_shifted);
}

Vector CfdSolver::compute_analytical_solution(Real t) const {
    Vector analytical(domain_.mesh().ncells());
    const auto& xcc = domain_.mesh().cell_centers();
    
    for (int i = 0; i < domain_.mesh().ncells(); ++i) {
        analytical[i] = analytical_solution(xcc[i], t);
    }
    return analytical;
}

void CfdSolver::write_solution(const std::string& filename, const Vector& solution) const {
    solution_.write_to_file(filename, solution, "CFD Solution");
}

void CfdSolver::write_current_solution(const std::string& filename) const {
    solution_.write_current_solution(filename, "CFD Current Solution");
}

void CfdSolver::print_info() const {
    std::cout << "\nCFD Solver Information:" << std::endl;
    std::cout << "==========================================" << std::endl;
    config_.print_brief();
    std::cout << "\n";
    domain_.print_info();
    std::cout << "  Time integrator: " << time_integrator_->name() << " (order " 
              << time_integrator_->order() << ")" << std::endl;
    std::cout << "  Residual calculator: " << residual_calculator_->name() << std::endl;
    std::cout << "  Boundary condition: " << boundary_condition_->name() << std::endl;
    std::cout << "  Initial condition: " << initial_condition_->name() << std::endl;
    std::cout << "==========================================" << std::endl;
}

// Static methods
Vector CfdSolver::run_single_simulation(const CfdConfig& config) {
    CfdSolver solver(config);
    return solver.run_simulation();
}

} // namespace cfd

File: example/1d-linear-convection/weno3/fortran/cfd/03h/cfd_solver.f90
! ==================== cfd_solver.f90 ====================
! OneFLOW-CFD Solver for 1D Convection Equation

module cfd_solver
    use kinds, only: dp
    use config_module, only: CfdConfigType
    use mesh_module, only: MeshType
    use domain_module, only: ComputationalDomainType
    use solution_module, only: SolutionType, update_oldfield
    use boundary_module, only: boundary
    use initial_condition_module, only: initial_condition, analytical_solution
	use recon_module, only: init_coef, ReconType,  reconstruction
    use time_integration_module, only: runge_kutta
    
    implicit none
    
    private

    ! ===================================================================
    ! Main CFD Solver Class
    ! ===================================================================
    type, public :: CfdType
        type(CfdConfigType), pointer :: config => null() 
        type(ComputationalDomainType) :: domain
        type(SolutionType) :: solution
        type(MeshType)  :: mesh
        type(ReconType), allocatable :: recon
    contains
        procedure :: init => cfd_init
        procedure :: init_field => init_field
        procedure :: run => run_simulation
    end type CfdType
    
    ! ===================================================================
    ! Public Procedures
    ! ===================================================================
    
    contains
    
    ! ===================================================================
    ! Initialization Methods
    ! ===================================================================
    
    ! CFD solver initialization
   
    subroutine cfd_init(this, config, mesh)
        class(CfdType), intent(inout) :: this
        type(CfdConfigType), intent(in), target :: config
        type(MeshType), intent(in) :: mesh
        
        this%config => config
        this%mesh = mesh
        call this%domain%init(config, mesh)
        
        call this%solution%init(this%domain)
		
        allocate(this%recon)
        call this%recon%init(config, this%domain)
        
        ! Adjust time step based on CFL condition
        call calculate_dt(this)
    end subroutine cfd_init    
    
    ! Calculate time step based on CFL condition
    subroutine calculate_dt(cfd)
        type(CfdType), intent(inout) :: cfd
        type(CfdConfigType), pointer :: config
        real(dp) :: dt_cfl
        
        config => cfd%config
        
        ! CFL condition: dt <= CFL * dx / |wave_speed|
        dt_cfl = config%cfl * cfd%domain%mesh%dx / abs(config%wave_speed)
        
        if (config%dt > dt_cfl) then
            print *, "Adjusting time step for stability:"
            print *, "  Original dt = ", config%dt
            print *, "  CFL dt = ", dt_cfl
            config%dt = dt_cfl
            print *, "  Using dt = ", config%dt
        end if

    end subroutine calculate_dt
    
    ! ===================================================================
    ! Field Initialization (使用外部模块的函数)
    ! ===================================================================
    
    ! Initialize field with step function
    subroutine init_field(this)
        class(CfdType), intent(inout) :: this
        integer :: i, j
        integer ::ist, ied
        ist = this%domain%ist
        ied = this%domain%ied
        
        do i = ist, ied - 1
            j = i - ist + 1
            this%solution%u(i) = initial_condition(this%domain%mesh%xcc(j))
        end do
        

        ! 使用 boundary_module 中的函数
        call boundary(this%solution%u, this%domain%nghosts, this%domain%ist, this%domain%ied)
        call update_oldfield(this%solution%un, this%solution%u, this%domain%ntcells)
    end subroutine init_field
    
     
    ! ===================================================================
    ! Simulation Driver
    ! ===================================================================
    
    ! Run simulation to final time
    function run_simulation(this) result(u_result)
        class(CfdType), intent(inout) :: this
        
        real(dp), allocatable :: u_result(:)
        
        real(dp) :: t, dt, dt_old, final_time
        integer :: step, max_steps
        
        type(CfdConfigType), pointer :: config
        
        
        allocate(u_result(this%domain%mesh%ncells))
        
        config => this%config
        final_time = config%final_time
        
        t = 0.0_dp        
        dt_old = config%dt
        dt = dt_old
        max_steps = 10000  ! Safety limit
        
        print *, "Starting time integration..."
        print *, "  Final time: ", final_time
        print *, "  Time step: ", dt
        print *, "  CFL number: ", config%cfl
        
        step = 0
        do while (t < final_time - 1.0e-12_dp .and. step < max_steps)
            step = step + 1
            
            if (t + dt > final_time) then
                dt = final_time - t
            end if
            
            config%dt = dt
            call runge_kutta(this%recon, config, this%domain, this%solution)
            t = t + dt
            
            ! Progress report
            if (mod(step, 100) == 0) then
                print *, "  Step ", step, ", Time = ", t
            end if
        end do
        
        if (step >= max_steps) then
            print *, "Warning: Reached maximum number of steps (", max_steps, ")"
        end if
        
        config%dt = dt_old
        
        print *, "Time integration completed:"
        print *, "  Total steps: ", step
        print *, "  Final time: ", t
        
        ! Extract physical solution (without ghost cells)
        u_result = this%solution%u(this%domain%ist:this%domain%ied)
    end function run_simulation
    
end module cfd_solver